### ISSUE TITLE
ENH: upstream tempita/cythonize wrappers

### ISSUE THREAD

**Issue by jbrockmendel** [cython/cython#2596]
We semi-recently started using `cythonize` in pandas, and I'm thinking some of the wrapping code used there might be worth pushing upstream.  I'd like to get opinions before making PR(s) here.

- Integrate optional tempita rendering: https://github.com/pandas-dev/pandas/blob/master/setup.py#L97

- Avoid running cythonize during `python setup.py clean`: https://github.com/pandas-dev/pandas/blob/master/setup.py#L463

- Enable cython coverage: https://github.com/pandas-dev/pandas/blob/master/setup.py#L442

**Comment by scoder** [comment]
I think enabling coverage/tracing support via an environment variable would be nice. There's always a certain risk that people might naively/accidentally export it globally and then get unexpected results from their next pip installation, but I guess that's just their fault then.

Regarding tempita templates, well, we ship the template engine, so making that a general feature seems like it could help some use cases. People have been requesting some kind of code generation support in the past, although that was also before fused types existed. Note that this would suggest some kind of original source line tracking, because error reporting on generated code can be confusing. Not sure if it's worth going down that road …

Finally, `clean` isn't the only command where compilation gets in the way. `setup.py --help-commands` lists quite a few, especially when `setuptools` is used. I think that basically brings us back to wanting better integration with setuptools in general.

**Comment by robertwb** [comment]
IIRC, the consensus on templating was that we should make it easy for users
to use a templating engine of their choice, rather than force one on the
users (or embed this as part of the language). Fused types did away with
much of the need for this, but still aren't a complete solution. As
mentioned, some way of tracking to the original line numbers should be
preserved.

Certainly worth writing up a proposal.

**Comment by jbrockmendel** [comment]
> Certainly worth writing up a proposal.

Great.  Is there a particular process for that?  Usually after discussing in an Issue I'd open a PR, but it sounds like that isn't what you have in mind.

**Comment by scoder** [comment]
I think the coverage envvar is ready for a PR, but the template engine idea deserves some more discussion first. We can continue it here and I'll rename the ticket.

Better build system support is being discussed in #2531.

**Comment by scoder** [comment]
Proposal:
- make Tempita available through a module-level Cython compiler option, e.g. `# cython: process_template=true`
- add a context manager as wrapper around `Utils.open_source_file()` which, if template processing was requested in the file, runs it through Tempita and returns the generated file as a file-like object (probably just a StringIO) instead of the original file
- change Tempita to generate `#::cython-line=123` markers to readjust the current source line counter, whenever the current line number differs from the expected line number (i.e. from incrementing the line numbers normally from the beginning or the last such marker)
- change `Lexicon.py` to handle `#::cython...` markers (at the beginning of a line) differently from other comments
- in `Scanners.py`, whenever such a marker is found, adapt `self.cur_line` accordingly
- optionally: add some way to set up a replacement context for Tempita, probably just externally and globally from `setup.py`, or as part of the `Extension` object.

(The `#::cython` marker was arbitrarily chosen as something that looks specific enough to avoid collisions with user comments. Other proposals welcome.)

We obviously then need a bunch of tests for this, especially tests for the line number reporting. Tests should go into `tests/build/`.

PR welcome.

**Comment by da-woods** [comment]
I'm basically in favour of this - I think it's a good idea to solve this kind of problem with something that works as a separate phase (and with distinct syntax) rather than trying to load too much into Cython syntax. Designing a robust "macro-like" language seems like a fairly hard problem, especially since Cython is pretty constrained to be close to Python syntax.

I've generally recommended Tempita for this kind of problem (just because it's there, and because people like Numpy and Scipy already use it), so building it into the build process a little more thoroughly seems like a good idea.

I'd prefer not killing off `DEF` and `IF` _too_ quickly. I definitely think we should deprecate them, issue a warning, and refuse to add new features to them. But I don't think they're causing much harm just by existing, so I'd rather leave a good long while for people to adapt and make sure that our proposed replacement is actually fit for purpose. To me Cython 3.1 seems too soon (given that the gap between 3.0 and 3.1 is currently planned to be fairly short).

**Comment by jakirkham** [comment]
Yeah some kind of templating in Cython could be pretty powerful, but probably needs to be done thoughtfully.

Tempita does seem like an easy way forward. Think there are some libraries already using it (like [scikit-learn]( https://github.com/scikit-learn/scikit-learn/search?q=tempita )). There are other templating libraries like this in the wild (like [Jinja]( https://jinja.palletsprojects.com )) that might also make sense. It might be worth considering how to templating while maintaining Python looking code (valuable for linting, reading by new devs, pure Python mode, etc.).

Fused types can already be pretty useful on their own. One can even use these to do [SFINAE]( https://en.cppreference.com/w/cpp/language/sfinae ) like things in Cython. For example here's an `enable_if` analogue in Cython ( https://github.com/cython/cython/issues/2660#issuecomment-687493653 ).

Have wondered if C++ style templating could be used in Cython directly.

Maybe it is worth enumerating what use cases templating would support?
